src/python/pants/backend/docker/rules.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+import logging
+import os
+from dataclasses import dataclass
+
+from pants.backend.docker.rules_binary import DockerBinary, DockerBinaryRequest
+from pants.backend.docker.rules_binary import rules as binary_rules
+from pants.backend.docker.rules_context import DockerBuildContext, DockerBuildContextRequest
+from pants.backend.docker.rules_context import rules as context_rules
+from pants.backend.docker.target_types import (
+    DockerContextRoot,
+    DockerImageSources,
+    DockerImageVersion,
+)
+from pants.core.goals.package import BuiltPackage, BuiltPackageArtifact, PackageFieldSet
+from pants.engine.fs import Digest, Snapshot
+from pants.engine.process import Process, ProcessResult
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import TransitiveTargets, TransitiveTargetsRequest
+from pants.engine.unions import UnionRule
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class DockerFieldSet(PackageFieldSet):
+    required_fields = (DockerImageSources,)
+
+    context_root_field: DockerContextRoot
+    image_version: DockerImageVersion
+    sources: DockerImageSources
+
+    @property
+    def dockerfile_name(self) -> str:
+        if not self.sources.value:
+            return "Dockerfile"
+        return self.sources.value[0]
+
+    @property
+    def dockerfile_path(self) -> str:
+        return os.path.join(self.address.spec_path, self.dockerfile_name)
+
+    @property
+    def context_root(self) -> str:
+        path = self.context_root_field.value or "."
+        if os.path.isabs(path):
+            path = os.path.relpath(path, "/")
+        else:
+            path = os.path.join(self.address.spec_path, path)
+        return path
+
+    @property
+    def image_tag(self) -> str:
+        return ":".join(s for s in [self.address.target_name, self.image_version.value] if s)
+
+
+@rule
+async def build_docker_image(
+    field_set: DockerFieldSet,
+) -> BuiltPackage:
+    docker = await Get(DockerBinary, DockerBinaryRequest())
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest([field_set.address]))
+    context = await Get(
+        DockerBuildContext,
+        DockerBuildContextRequest(
+            field_set.address, field_set.context_root, tuple(transitive_targets.closure)
+        ),
+    )
+
+    snapshot = await Get(Snapshot, Digest, context.digest)
+    files = "\n".join(snapshot.files)
+    logger.debug(
```

Comment:
Is this for your own debugging, or should this stay in for the end user? `files` could be huge, no?

Reply:
Not that huge, but as it stands, I'd say for debugging only, now. However I think it would be a great aid to have some way as end user to introspect what the build context contains.. (maybe the preserve cache dirs option thingy is enough for that?)